Commercial/custom pricing (#690)
shopware/docs: guides/plugins/plugins/api/custom-pricing.md

# Custom Pricing

The Custom Pricing feature allows massive advances in the pricing model capabilities in the Shopware 6 ecosystem.

The API interface exposed by this module allows the user to operate a set of commands which will enable the granular 
''overriding'' of prices via an external data repository or ERP system. This is achieved by defining a custom 

> relationship between a current price and one of the following entities Customer.

Author: ```suggestion
relationship between the current price and one of the following entities:
- Customer
```

### Pre-requisites and setup

As Custom Pricing is part of the Commercial plugin, it requires an existing Shopware 6 installation and the activated
Shopware 6 Commercial plugin on top. This plugin Commercial plugin can be installed as per the familiar

> [install instructions](guides/plugins/plugins/plugin-base-guide#install-your-plugin). In addition, Custom Prices need 

Author: ```suggestion
[install instructions](guides/plugins/plugins/plugin-base-guide#install-your-plugin). In addition, the `Custom Prices` feature needs 
```

> to be activated within your merchant's account.

Author: ```suggestion
to be activated within the relevant merchant account.
```

## Working with the API route

> To create, alter and delete customer-specific prices, you can use the API endpoint `/api/_action/product-customer-price`. As like with any other admin request in Shopware, you first need to take care of authenticating yourself. Therefore, please head over to the 

Author: ```suggestion
To create, alter and/or delete customer-specific prices, you can use the API endpoint `/api/_action/product-customer-price`. As like with any other admin request in Shopware, you first need to authenticate yourself. Therefore, please head over to the 
```

[authentication guide](https://shopware.stoplight.io/docs/admin-api/ZG9jOjEwODA3NjQx-authentication) for details.

> Apart from that, the Custom Pricing API interface models itself upon the interface of the sync API, so you will 

Author: ```suggestion
Otherwise, the Custom Pricing API interface models itself upon the interface of the sync API, so you will 
```

be able to package your requests similarly, see our [API documentation](https://shopware.stoplight.io/docs/admin-api).

{% hint style="info" %}

> You can use the route with single `upsert` and `delete` actions or even combine those in a single request. This way, you can pack several different commands inside one sync API request, and each of them would be executed independent and isolate

Author: ```suggestion
You can use the route with single `upsert` and `delete` actions or even combine those in a single request: you can pack several different commands inside one sync API request, and each of them is executed in an independent and isolated way
```

{% endhint %}

So, it's not surprising that the request body looks like a familiar sync request. In the payload for the `upsert` action, you pass the following data:
* `productId`: The product whose price should be overwritten

> * `customerId`: The customer for whom we'd like to have a custom price for

Author: ```suggestion
* `customerId`: The customer for whom we will assign a custom price
```

> * `price`: The new custom price you want to use. It's defined by how you're familiar when working with prices in Shopware.

Author: ```suggestion
* `price`: The new custom price you want to use.
```
Not actually sure this is correct, we have included the quantityStart and quantityEnd within the payload, and then there are sub-prices with the same schema as the normal price json from `product_price`

This way, we come to use a payload as seen in the example below:
```json
[
  {
    "action": "upsert",
    "payload": [
      {
        "productId": "0001e32041ac451386bf9b7351c540f3",
        "customerId": "02a3c82b5ca842c492f8656029b2e63e",
        "price": [
          {
            "quantityStart": 1,
            "quantityEnd": 10,
            "price": [
              {
                "currencyId": "b7d2554b0ce847cd82f3ac9bd1c0dfca",
                "gross": 682.0,
                "net": 682.0,
                "linked": true
              }
            ]
          }
        ]
      }
    ]
  }
]
```

> For the `delete` action, we need fewer data to send. For its payload, you only need to define the `productId`s of the product whose prices you want to remove.

Author: ```suggestion
For the `delete` action, we will send a smaller dataset: you only need to define the `productId`s for the product whose prices you want to remove.
```

```json
[
  {
    "action": "delete",
    "payload": [
      {
        "productIds": [
          "0001e32041ac451386bf9b7351c540f3",
          "363a6985f6434a7493b1ef3dabeed40f"
        ]
      }
    ]
  }
]
```

{% hint style="info" %}
In case of an error occurs, the response will not return an error code - which is typical for the sync API; instead, any validation errors will be stored within the `errors` key. 
{% endhint %}

{% hint style="warning" %}

> When working with this route, one difference sets it apart from familiar `sync` requests: You cannot specify headers to adapt the endpoint's behavior. 

Author: ```suggestion
When working with this route, one difference sets it apart from the familiar `sync` requests: You cannot specify headers to adapt the endpoint's behavior. 
```

> # Custom Pricing

Reviewer: Please add a corresponding entry in the `SUMMARY.md` file for both new pages in the PR.

Otherwise, the Custom Pricing API interface models itself upon the interface of the sync API, so you will 
be able to package your requests similarly, see our [API documentation](https://shopware.stoplight.io/docs/admin-api).

{% hint style="info" %}
You can use the route with single `upsert` and `delete` actions or even combine those in a single request: you can pack several different commands inside one sync API request, and each of them is executed in an independent and isolated way
{% endhint %}

So, it's not surprising that the request body looks like a familiar sync request. In the payload for the `upsert` action, you pass the following data:
* `productId`: The product whose price should be overwritten
* `customerId`: The customer for whom we will assign a custom price
* `price`: The new custom price you want to use.

This way, we come to use a payload as seen in the example below:
```json
[
  {
    "action": "upsert",
    "payload": [
      {
        "productId": "0001e32041ac451386bf9b7351c540f3",
        "customerId": "02a3c82b5ca842c492f8656029b2e63e",
        "price": [
          {
            "quantityStart": 1,

>             "quantityEnd": 10,

Reviewer 2: This value makes the prices look weird, use `null` instead

Reviewer 3: @elkmod - can this be a null ? as it must be of type integer.

Author: This is already outdated, please see new code

Author: @sushmangupta I recently updated the spotlight and Commercial plugin API definition, `null` is accepted
https://github.com/shopware/admin-api-reference/pull/2/files#diff-af2ce1e09fc283bbb187fdacd0e82764557cbe3e8141e5e4d79beda4a080656aR11762
